### PR TITLE
test: add GDP hybrid dynamic-inference functional tests (GB200 + H100)

### PR DESCRIPTION
**Pull request by shanmugamr1992**
## Summary

Adds functional tests for **Gated Delta Product (GDP)** hybrid dynamic inference on **both GB200 and H100**:

- `tests/functional_tests/test_cases/hybrid/hybrid_gdp_dynamic_inference_tp1_pp1_dp4_800m/` (GB200) — `model_config.yaml` + `golden_values_dev_dgx_gb200.json`
- `tests/functional_tests/test_cases/hybrid/hybrid_gdp_dynamic_inference_tp1_pp1_dp8_800m/` (H100) — `model_config.yaml` + `golden_values_dev_dgx_h100.json`
- `tests/test_utils/recipes/gb200/hybrid-dynamic-inference.yaml` — new recipe for the GB200 case
- `tests/test_utils/recipes/h100/mamba-dynamic-inference.yaml` — H100 case registered on the existing recipe (it already hosts every `hybrid_dynamic_inference_*` case and has `model: hybrid`)

Both drive `examples/inference/advanced/gpt_dynamic_inference.py` on the GDP hybrid model and compare `generated_tokens` + `logprobs` against recorded golden values.

The two cases differ only in data parallelism, which follows the GPUs-per-node of each recipe: the GB200 recipe exports `GPUS_PER_NODE=4`, while the H100 recipe leaves `_run_training.sh`'s default of 8, so the cases are dp4 and dp8 respectively. Model config, sampling, and prompt are otherwise identical.

## Synthetic checkpoint note (model + where it lives)

No public GDP checkpoint exists, so a **synthetic** one was created for this test.

- **Model:** synthetic GDP (Gated Delta Product) hybrid, ~880M params, `bf16`, `torch_dist` (DCP) format. Built with `pretrain_hybrid.py` using `gdp_stack_spec` (`megatron.core.models.hybrid.hybrid_layer_specs`). Config: hidden 1024, ffn 4096 (swiglu), 16 attention heads / 8 GQA groups, 32 mamba heads. ~12 GB on disk (includes optimizer state; the tests load with `--no-load-optim`).
- **Clusters:** the `mcore_ci` artifacts tree is replicated per-cluster, so the checkpoint has been mirrored to **both** clusters that run these tests — **OCI-hsg** (GB200 / Blackwell) and **cw-dfw** (H100 / Hopper). The same relative path and identical bytes are present on both.
- **Checkpoint path (identical on both clusters):**
  `/lustre/fsw/portfolios/coreai/projects/coreai_dlalgo_mcore/mcore_ci/model/gdp_hybrid_800m/dcp/mcore-v1_bf16/checkpoint`
- **Tokenizer (TikTokenizer) alongside it:**
  `/lustre/fsw/portfolios/coreai/projects/coreai_dlalgo_mcore/mcore_ci/model/gdp_hybrid_800m/dcp/mcore-v1_bf16/multiMixV8.gpt4o_nc_sd.500000.128k.vocab.json`

Any additional cluster that picks up these tests will need the same mirror.

## Verification

**GB200 (OCI-hsg, job 6131506):** checkpoint loads, dynamic inference runs deterministically, golden comparison passes:

```
Running pytest checks against golden values
test_inference_regular_pipeline.py::test_inference_pipeline PASSED
============================== 1 passed in 0.16s ===============================
```

**H100 (cw-dfw, job 15962800):** goldens were bootstrapped in lightweight mode (job 15961641), then a fresh run was compared against them and matched bit-for-bit:

```
Running pytest checks against golden values
test_inference_regular_pipeline.py::test_inference_pipeline PASSED
============================== 1 passed in 0.13s ===============================
```

## Notes

- Commits are DCO signed-off (`-s`) but **not GPG-signed** — `gpg` is unavailable in the authoring environment. Will re-sign/amend with `-S` before merge if required for copy-pr-bot / `/ok to test`.
- Both cases are scoped `[mr, mr-github]`, so they run in the internal GitLab matrix and in the GitHub PR pipeline / merge-group gate. This matches the sibling `hybrid_dynamic_inference_*` cases in the h100 recipe.

## Test plan

- [ ] `Run functional tests` label (`scope=mr-github`, `n_repeat=5`, golden-value compare) picks up both new cases
- [ ] `hybrid_gdp_dynamic_inference_tp1_pp1_dp4_800m` passes golden comparison in CI (GB200)
- [ ] `hybrid_gdp_dynamic_inference_tp1_pp1_dp8_800m` passes golden comparison in CI (H100)

### CODE REVIEW

**Comment by shanmugamr1992** [comment]
/ok to test [d4b85a8](https://github.com/NVIDIA/Megatron-LM/pull/6526/commits/d4b85a89940e423b1b4a330472bdb34a9ef715a9)

**Comment by shanmugamr1992** [comment]
/ok to test [48ebbf0](https://github.com/NVIDIA/Megatron-LM/pull/6526/commits/48ebbf0ddfad39ae51f9434bece94c43a52b5797)

**Comment by shanmugamr1992** [comment]
### Correction to the root cause in commit `6fe9992`

The commit message for `6fe9992` attributes the GB200 golden mismatch to the OCI-hsg copy of the synthetic `gdp_hybrid_800m` checkpoint holding different weights. **That is incorrect** — the checkpoint is fine on both clusters. Recording the corrected analysis here so the reasoning in the history isn't misleading.

That conclusion was drawn from the Pearson correlation (0.65) between the committed GB200 golden and the CI output across prompt logprobs, which is not a discriminating statistic. The discriminating one is **prompt position 0**, where the SSM recurrent state is still empty and the forward pass is effectively feed-forward:

| | `prompt_logprobs[0]` |
|---|---|
| GB200 golden (recorded on OCI-hsg) | −11.472631 |
| GB200 CI output | −11.473866 |
| H100 golden | −11.473866 |

Agreement to 0.0012 nats rules out independently-initialized weights. Distributions match as well (mean −12.098 vs −12.067, std 0.627 vs 0.644), and the divergence does not accumulate smoothly — it is 0.001 at position 0, then jumps to 0.64 at position 1 and scatters in the 0–1.4 range.

**Actual cause:** same weights, different GDP/SSM kernel build. `mamba-ssm` / `causal-conv1d` live in the `ssm` extra and are not in the cog env, so they are hand-built per cluster (the `.so` files are arch-specific), while CI uses the build shipped in the `mcore_ci_dev` image. Position 0 has no recurrence and agrees to 1e-3; from position 1 the recurrent path diverges. Because the checkpoint is untrained — mean generated logprob −9.16 against a uniform baseline of −11.78 over a 131072 vocab — the top-1/top-2 margin is tiny, so that is enough to flip argmax and rewrite all 30 generated tokens.

The change itself is unaffected: the goldens were re-recorded from the CI run, which is the environment that gates this PR. Supporting evidence for that — the GB200 CI output's prompt logprobs are bit-identical to the H100 golden, which does reproduce in CI.

**Worth noting for reviewers:** these goldens are pinned to the CI container's kernel build. A base-image bump that moves `mamba-ssm` or `causal-conv1d` will break them again. The underlying fragility is exact-token-match on an untrained checkpoint; a trained checkpoint or a tolerance-based metric would make these tests durable, likely as a follow-up rather than a blocker here.

**Review by chtruong814 — APPROVED** [review]

**Comment by shanmugamr1992** [comment]
/ok to test [23a0196](https://github.com/NVIDIA/Megatron-LM/pull/6526/commits/23a0196d07b35ac1d25ae155e5516139cfd1ef23)

**Comment by shanmugamr1992** [comment]
/ok to test [fbad0ba](https://github.com/NVIDIA/Megatron-LM/pull/6526/commits/fbad0bab6923ad355289d9d9b29ac0864e0e5800)

**Review by santhnm2 — APPROVED** [review]

**Review by ericharper — APPROVED** [review]

**Review by yaoyu-33 — APPROVED** [review]

**Review by jaredcasper — APPROVED** [review]

**Comment by shanmugamr1992** [comment]
/ok to test [81bc5cc](https://github.com/NVIDIA/Megatron-LM/pull/6526/commits/81bc5cc5054307e95d16091920bd4f937a42a413)

**Comment by shanmugamr1992** [comment]
/ok to test [4372021](https://github.com/NVIDIA/Megatron-LM/pull/6526/commits/43720214cd295846b4c154ca10ef41f084aa5b31)
